Add a 1em bottom margin below the save button

kaapstorm opened the pull request:
Hey @biyeun, I want to add a 1em margin below a button, otherwise it touches a well below it and looks dumb.

But I'm not sure where my class definition should go. I'm sure "utilities.less" is probably the wrong place, but the other places all looked even more wrong.

Also unsure about class naming conventions, if any; "one-em-bottom" would be a bad name if we decided that 10px would be better.

Could you tell me what I should change, or point me to a style guide?

Buddy @millerdev

czue committed:
allow overriding form_extras on submission + tests

czue committed:
add test and fix bug

czue committed:
replace method with case factory

Commit:
Added Functionality Checklist report to Asha report

Commit:
code refactor

Commit:
added subtasks in stock data migration

Commit:
Tweaks in stock data task

proteusvacuum committed:
rebuild cases when users are deleted

proteusvacuum committed:
offerings to dimagimon

proteusvacuum committed:
No cheating in tests + code cleanup

proteusvacuum committed:
Gifts to dimagimon

proteusvacuum committed:
update set instead of looping + adding

benrudolph committed:
Don't swallow error

benrudolph committed:
Add semicolon

agrzywinski committed:
ILSGateway: Add MSD zones

czue committed:
don't rename advanced modules for commtrack apps

Commit:
Refactored stock data task

Commit:
Merge branch 'master' into ils-migration

kaapstorm committed:
Import just the tests you want to run

otherwise base test classes will be run too.

Commit:
flake8

kaapstorm committed:
Merge branch 'master' into import_specific_tests

Conflicts:
	corehq/apps/products/tests/__init__.py

agrzywinski committed:
Capitalize text in export button

czue committed:
Merge pull request #6608 from lwyszomi/166646-2

Capitalize text in export button

czue committed:
add admin file for ownership cleanliness

agrzywinski committed:
EWS: increase timeout in one scirpt

Commit:
changes in daterange filer

Commit:
flake

snopoke committed:
support using mod_xsendfile and x-accel to download files

kaapstorm committed:
Append "Z" to Program datetimes if necessary

Commit:
Use subtasks only during initial migration

czue committed:
add function to set cleanliness flags + tests

czue committed:
add get_indexed_cases and refactor functions

czue committed:
Merge pull request #6612 from dimagi/bug/167857/program_datetime

Append "Z" to Program datetimes if necessary

czue committed:
fix bug in get_indexed cases and use domain and ids instead of full cases

this undoes some of the refactoring from the last commit since
it was broken

czue committed:
add hint validation logic and tests

snopoke committed:
Merge pull request #6610 from dimagi/ownership-cleanliness

add admin file for ownership cleanliness

snopoke committed:
Merge pull request #6607 from dimagi/import_specific_tests

Import just the tests you want to run

czue committed:
delete unused function

snopoke committed:
Merge pull request #6404 from lwyszomi/up-nrhm-reports

Up nrhm reports

czue committed:
add management command to set cleanliness flags for one or all domains

czue committed:
rename model and remove primary key constraint for owner_id

czue committed:
add initial south migration

czue committed:
always update last time checked

czue committed:
properly handle cleanliness with owners spanning multiple domains

czue committed:
missed migration file

czue committed:
fix admin references

kaapstorm committed:
Add a 1em bottom margin below the save button

otherwise it touches the well below it, and looks dumb.

czue commented:
fyi @biyeun is on vacation for two weeks so maybe @dannyroberts or @orangejenny might know?

@kaapstorm style guid is here: https://www.commcarehq.org/styleguide/ but i don't think it answers your question anywhere. 

(this change isn't obviously wrong to me, but I always double check with @biyeun sadly)

biyeun commented on corehq/apps/style/static/style/lib/bootstrap/less/utilities.less:
```diff
@@ -28,3 +28,8 @@
 }
+
+// TODO: Move this to wherever it belongs
+.one-em-bottom {
```
this isn't the right place. `/static/style/less` is where all the less updates go `lib` should not be touched.

dannyroberts committed:
Merge pull request #6614 from dimagi/dead-code

delete unused function

benrudolph commented:
What about using existing classes we have? it looks like that div has no styles applied to it? not sure exactly what that ko binding adds to the div but maybe try `class="btn btn-primary"`.

kaapstorm commented:
Thanks @czue. Sorry @biyeun, forgot your vac started now. Enjoy your leave. Thanks @benrudolph I'll check those out.

benrudolph committed:
Log the actual error message to the console

czue committed:
[ci skip] use dbaccessors

kaapstorm commented:
@benrudolph, SaveButton is defined [here][1]. It uses existing Bootstrap button styles, which are defined [here][2]. To add a bottom margin, I think I should add a new class, because I couldn't find a Bootstrap class that does that. I just don't know where to define it -- somewhere under corehq/apps/style/static/style/less -- but where?

... Actually, it seems appstore just puts the style [inline][3]. I'm a little ambivalent about that.

  [1]: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/hqwebapp/static/hqwebapp/js/main.js#L31
  [2]: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/style/static/style/lib/bootstrap/docs/assets/css/bootstrap.css#L3149
  [3]: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/appstore/templates/appstore/partials/sidebar.html#L15

czue committed:
make dbaccessors a package

dannyroberts commented:
Wish I had a stronger opinion, but I'm fine with anything that doesn't edit style/lib like biyeun said

kaapstorm committed:
Add top and bottom layout classes

Add .top* and .bottom* classes for layout. Put
lib/bootstrap/less/utilities.less back the way it was.

czue committed:
move functions to dbaccessors

czue committed:
more stuff to dbaccessors

kaapstorm commented on corehq/apps/style/static/style/less/components/layouts.less:
```diff
@@ -87,3 +87,35 @@
+
+.top1em {
+  margin-top: 1em;
+}
```
This feels better to me; it allows other people to use the same classes for layout, it seems like a better place to put these and creates a place to add more if, for some reason, you need a 5px margin, or whatever.

dannyroberts committed:
add request object to soft_assert when possible

orangejenny commented:
Apologies for length.

You might want to be applying a style to a parent element, not to the button itself. If there were text next to the button, would it also need a bottom margin so as not to crowd the well? If so, then you should add padding/margin to the `span12` or the `row-fluid` or the containing div, not to the button itself. I'm not sure what the overall page markup looks like here, but seems like it might be a case for styling your container element.

On the other hand (there are so many options!)...this is a partial. Is it always going to be directly above other content? It's possible that instead of styling the button or one of its ancestors, the markup where this template gets included should put some space between usercase_config.html and the content below it. This could be an empty paragraph, could be a blank row, could be a horizontal rule.

In general, styles are better when they're semantic. The rough thing about classes like `top10px` and `top1em` (besides that they're different systems of units) is that they're meaningless. (This is one of the reasons I'm with you on inline styles being sketchy.) One dev will slap on a `top10px` and another will grab the `top20px` and you end up with a site that has a different amounts of "breathing room" in different areas, so it looks just a little bit less than professional but it's hard to put your finger on why. Ideally, we get to a place where we have standard container classes and the person marking up content for a page never has to deal with content bumping up against other content, and they're definitely not deciding between 10px and 20px. At the moment, I'd probably either add a top/bottom margin to the row-fluid (possibly to all row-fluids on the page where this appears), or maybe just put an empty paragraph below the .row-fluid and call it a day.

As an aside, it would be nice if it were more obvious whether to modify /static/style/less or /static/style/lib, I've gotten tripped up on that, too. Maybe a one-line README in both directories?

czue committed:
fix test failure output

snopoke committed:
Merge pull request #6620 from dimagi/test-fail-output

fix test failure output

emord committed:
Merge pull request #6606 from dimagi/rename

don't rename advanced modules for commtrack apps

benrudolph commented:
:heart: everything you just said

snopoke committed:
Merge pull request #6616 from dimagi/cleanliness-setter

add management command that can set cleanliness flags

dannyroberts committed:
split soft_assert message from specific value

and exclude specific value from subject
for email threading purposes

TylerSheffels committed:
Updating submodules

[ci skip]

dannyroberts committed:
Merge pull request #6618 from dimagi/dbaccessors

move to dbaccessors

emord committed:
Merge pull request #6621 from dimagi/case-factory-tweaks

Case factory tweaks

benrudolph committed:
Merge pull request #6619 from dimagi/soft-assert-request

add request object to soft_assert when possible

dannyroberts committed:
fix import error

dannyroberts committed:
random cleanup of case/dbaccessors/

while I'm here

czue commented:
:+1: especially on the one line README!

czue committed:
Merge pull request #6622 from dimagi/cleanliness-setter

dbaccessors

czue committed:
fix function

NoahCarnahan committed:
Merge pull request #6624 from dimagi/fix

fix function

czue committed:
Merge pull request #6604 from dimagi/use-better-backbone-logging

Don't swallow error

czue committed:
don't set instance properties if not necessary

czue committed:
move sync token association to factory default

czue committed:
switch test to factory

czue committed:
replace case creation with factory

czue committed:
replace methods with factory

czue committed:
add demo_user_group_id to list of weird user ids

czue committed:
add tests for cross-domain cleanliness

czue committed:
fix hint lookup and change clean flag to hard failure

czue committed:
don't set hints for weird user ids

dannyroberts committed:
have TestGatewayFee use setUpClass

takes 30 seconds off tests

dannyroberts committed:
quickcache lookup of sms backend is_global for testing speed

dannyroberts committed:
have TestGatewayFee bulk delete in tearDownClass

dannyroberts committed:
Merge pull request #6625 from dimagi/case-factory

minor cleanup to sync mode tests

dannyroberts committed:
rebalance travis job split

czue committed:
Merge pull request #6613 from lwyszomi/ils-migration

Changes in stock data migration

czue committed:
Merge pull request #6609 from lwyszomi/167135-2

EWS: increase timeout in one scirpt

czue committed:
Merge pull request #6605 from lwyszomi/166406

ILSGateway: Add MSD zones

proteusvacuum committed:
dont use mutable type as default arg

czue committed:
start splitting out concerns in the restore code

czue committed:
add TemporaryRestoreConfig class to help migrate to new API

Commit:
Include only products assigned to location in reminder

czue committed:
remove unused member

czue committed:
add caching params to the TemporaryRestoreConfig API

snopoke committed:
update staging.yaml

[ci skip]

snopoke committed:
migrate dev environment

czue committed:
update commtrack tests to use new API, rely on domain stock_settings

czue committed:
remove stock_settings param from restore config

czue committed:
make remove TemporaryRestoreConfig, replace constructor signature

czue committed:
[ci skip] lint

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
Merge branch 'usercase' into save_button_bottom_margin

czue committed:
move restore-related exceptions to the phone module

czue committed:
remove dead comment pointing to missing comment

czue committed:
clean up validation a bit - push to state object and raise better error

czue committed:
[ci skip] lint

sravfeyn committed:
Merge pull request #6629 from dimagi/should-migrate

migrate

kaapstorm committed:
Styling based on Jenny's feedback

czue committed:
remove items

czue committed:
unused import

Commit:
Bugfix - added property decorator to method

czue committed:
whitespace

czue committed:
push more sync token logic into the state object. rename things a bit

czue committed:
switch all non-case/ledger restore components to use data providers

czue committed:
[ci skip] lint

dannyroberts committed:
remove unused code from dotsSubmissionTests

Commit:
Add check in web user sync

czue committed:
add domain and provider_log to RestoreState

czue committed:
push restore_class into the state object

snopoke committed:
Merge pull request #6626 from dimagi/ownership-cleanliness

Ownership cleanliness fixes and tweaks

czue committed:
introduce long running provider, convert case payload to provider

czue committed:
[ci skip] unused variable

czue committed:
Merge pull request #6631 from lwyszomi/ilsgateway-hotfix

Bugfix - added property decorator to method

czue committed:
Merge pull request #6627 from dimagi/test_gateway_fees-setUpClass

speed up TestGatewayFee

czue committed:
Merge pull request #6632 from dimagi/pact-test-cleanup

remove unused code from dotsSubmissionTests

czue committed:
Merge pull request #6628 from lwyszomi/166660-case

Include only products assigned to location in reminder

millerdev committed:
vellum:eff463e5dec88619ffa1d6a27c830cc1054cbad3

benrudolph committed:
Merge pull request #6615 from dimagi/django-transfer

support using mod_xsendfile and x-accel to download files

esoergel committed:
Merge pull request #6584 from dimagi/deleted-user-spinner

rebuild re-assigned cases when users are deleted

snopoke committed:
Merge pull request #6630 from dimagi/clean-restore

refactoring restore code a bit

emord committed:
vellum:087b035cc34ab9ae16d676bf05afd3d5e6310028

kaapstorm committed:
Increase LateFlag.template_width to 11%

millerdev committed:
Merge pull request #6642 from dimagi/bug/167709/late_flag_width

Increase LateFlag.template_width to 11%

kaapstorm committed:
Added README file for custom styles

dimagimon commented on corehq/apps/locations/tests/__init__.py:
```diff
@@ -1,6 +1,11 @@
 from .test_location_import import *
 from .test_location_types import *
-from .test_locations import *
+from .test_locations import (
```
F401 'LocationGroupTest' imported but unused

dimagimon replied:
F401 'LocationProducts' imported but unused

dimagimon replied:
F401 'LocationsTest' imported but unused

dimagimon replied:
F401 'WrapLocationTest' imported but unused

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
Merge branch 'usercase' into save_button_bottom_margin

dimagimon commented:
There are 3 errors:

* settings.py, line 501 - E265 block comment should start with '# '
* settings.py, line 508 - E265 block comment should start with '# '
* settings.py, line 968 - E265 block comment should start with '# '

biyeun commented:
everything is static/style/lib is a library and should remain untouched.

On Fri, May 8, 2015 at 11:19 AM, Jenny Schweers <notifications@github.com>
wrote:

> Apologies for length.
>
> You might want to be applying a style to a parent element, not to the
> button itself. If there were text next to the button, would it also need a
> bottom margin so as not to crowd the well? If so, then you should add
> padding/margin to the span12 or the row-fluid or the containing div, not
> to the button itself. I'm not sure what the overall page markup looks like
> here, but seems like it might be a case for styling your container element.
>
> On the other hand (there are so many options!)...this is a partial. Is it
> always going to be directly above other content? It's possible that instead
> of styling the button or one of its ancestors, the markup where this
> template gets included should put some space between usercase_config.html
> and the content below it. This could be an empty paragraph, could be a
> blank row, could be a horizontal rule.
>
> In general, styles are better when they're semantic. The rough thing about
> classes like top10px and top1em (besides that they're different systems
> of units) is that they're meaningless. (This is one of the reasons I'm with
> you on inline styles being sketchy.) One dev will slap on a top10px and
> another will grab the top20px and you end up with a site that has a
> different amounts of "breathing room" in different areas, so it looks just
> a little bit less than professional but it's hard to put your finger on
> why. Ideally, we get to a place where we have standard container classes
> and the person marking up content for a page never has to deal with content
> bumping up against other content, and they're definitely not deciding
> between 10px and 20px. At the moment, I'd probably either add a top/bottom
> margin to the row-fluid (possibly to all row-fluids on the page where this
> appears), or maybe just put an empty paragraph below the .row-fl uid and
> call it a day.
>
> As an aside, it would be nice if it were more obvious whether to modify
> /static/style/less or /static/style/lib, I've gotten tripped up on that,
> too. Maybe a one-line README in both directories?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/6617#issuecomment-100270405>.
>

biyeun commented:
I'd like to echo what @orangejenny said, that the margin should be applied to the parent block element, not the button.

kaapstorm commented:
Thanks @biyeun, I agree. I took @orangejenny's advice, added a short README to static/style/lib, and applied the style at the block level. I also dropped the range of styles I had created for different sized gaps, and replaced them with just two, "top-spacer" and "bottom-spacer", with an explanation for why I made them 9px. It's in [commit 3c3c4bd9](https://github.com/dimagi/commcare-hq/commit/3c3c4bd9d4709bf938968098a54c5c0725831ab8). Please let me know if you think I should change anything.
